src/color-retriever/index.ts

```diff
@@ -0,0 +1,194 @@
+import { Extracted } from "../extractor";
+import { QRLocation } from "../locator";
+
+export interface QRColors {
+  qr: RGBColor;
+  background: RGBColor;
+}
+
+export interface RGBColor {
+  r: number;
+  g: number;
+  b: number;
+}
+
+interface CIELabColor {
+  L: number;
+  a: number;
+  b: number;
+}
+
+/* Retrieves the colors that make up a scanned QR code. RGB (assumed to be sRGB) values are converted to the CIELab
+color space for averaging (with no regard for alpha), and then converted back to RGB. */
+export function retrieveColors(location: QRLocation,
+                               extracted: Extracted,
+                               sourceData: Uint8ClampedArray,
+                               sourceWidth: number): QRColors {
+
+  const qrColor: CIELabColor = {L: 0, a: 0, b: 0};
+  const backgroundColor: CIELabColor = {L: 0, a: 0, b: 0};
+  let qrPixels = 0;
+  let backgroundPixels = 0;
+
+  for (let y = 0; y < location.dimension; y++) {
+    for (let x = 0; x < location.dimension; x++) {
+      const sourcePixel = extracted.mappingFunction(x + 0.5, y + 0.5);
+      const sourcePixelOffset = ((Math.floor(sourcePixel.y) * sourceWidth) + Math.floor(sourcePixel.x)) * 4;
+      const sourceColor = rgbToLab({r: sourceData[sourcePixelOffset],
+                                          g: sourceData[sourcePixelOffset + 1],
+                                          b: sourceData[sourcePixelOffset + 2]});
+
+      if (extracted.matrix.get(x, y)) {
+        qrColor.L += sourceColor.L;
+        qrColor.a += sourceColor.a;
+        qrColor.b += sourceColor.b;
+        qrPixels++;
+      } else {
+        backgroundColor.L += sourceColor.L;
+        backgroundColor.a += sourceColor.a;
+        backgroundColor.b += sourceColor.b;
+        backgroundPixels++;
+      }
+    }
+  }
+
+  qrColor.L /= qrPixels;
+  qrColor.a /= qrPixels;
+  qrColor.b /= qrPixels;
+  backgroundColor.L /= backgroundPixels;
+  backgroundColor.a /= backgroundPixels;
+  backgroundColor.b /= backgroundPixels;
+
+  return {
+    qr: labToRGB(qrColor),
+    background: labToRGB(backgroundColor),
+  };
+}
+
+// Color space conversions from http://www.easyrgb.com/en/math.php
+
+// Converts an RGB color ([r, g, b] or [r, g, b, a] - a is ignored) to CIELab ([L*, a*, b*]).
```

Comment:
This comment isn't strictly correct anymore I don't think - Can probably just remove the array descriptions from it?